crates/kitsune_p2p/kitsune_p2p/src/gossip/simple_bloom/step_2_local_sync_inner.rs

```diff
@@ -0,0 +1,224 @@
+                                    .gossip(
+                                        space.clone(),
+                                        new_agent.clone(),
+                                        old_agent.clone(),
+                                        key.clone(),
+                                        data.clone(),
+                                    )
+                                    .await
+                                    .map_err(KitsuneError::other)?;
+                            }
+                            // this should be impossible right now
+                            // due to the shared agent store
+                            MetaOpData::Agent(_) => unreachable!(),
+                        }
+
+                        new_set.insert(old_key.clone());
+                    }
+                }
+            }
+        }
+
+        if local_synced_ops > 0 {
+            tracing::debug!(
+                %local_synced_ops,
+                "local sync",
+            );
+        }
+
+        *has_hash = new_has_map;
+
+        Ok(())
+    }
+
+    pub fn finish(self) -> (DataMap, KeySet, BloomFilter) {
+        let Self {
+            data_map, has_hash, ..
+        } = self;
+
+        // 1 in 100 false positives...
+        // we can get 1 in 1000 for ~2x the filter size, but may not be worth it
+        // 1 in 100 pretty much guarantees full sync after two communications.
+        const TGT_FP: f64 = 0.01;
+
+        // at this point, all the local has_hash maps should be identical,
+        // so we can just take the first one
+        let (key_set, bloom) = if let Some((_, map)) = has_hash.into_iter().next() {
+            let len = map.len();
+            tracing::trace!(
+                local_op_count=%len,
+                "generating local bloom",
+            );
+            let mut bloom = bloomfilter::Bloom::new_for_fp_rate(len, TGT_FP);
+            for h in map.iter() {
+                bloom.set(h);
+            }
+            (map, bloom)
+        } else {
+            (HashSet::new(), bloomfilter::Bloom::new(1, 1))
+        };
+
+        (data_map, key_set, bloom)
+    }
+}
+
+async fn data_map_get(
+    evt_sender: &mut futures::channel::mpsc::Sender<event::KitsuneP2pEvent>,
+    space: &Arc<KitsuneSpace>,
+    agent: &Arc<KitsuneAgent>,
+    map: &mut DataMap,
+    key: &Arc<MetaOpKey>,
+) -> KitsuneResult<Arc<MetaOpData>> {
+    use crate::event::*;
+    if let Some(data) = map.get(key) {
+        return Ok(data.clone());
+    }
+    match &**key {
+        MetaOpKey::Op(key) => {
+            let mut op = evt_sender
+                .fetch_op_hash_data(FetchOpHashDataEvt {
+                    space: space.clone(),
+                    agent: agent.clone(),
+                    op_hashes: vec![key.clone()],
+                })
+                .await
+                .map_err(KitsuneError::other)?;
+
+            if op.len() != 1 {
+                return Err("invalid results".into());
```

Comment:
This error message is going to be hard to understand out of context

Reply:
:+1: fixed